### ISSUE TITLE
Symlinks mangled in 1.3.6

### ISSUE THREAD

**Issue by kdkd** [proftpd/proftpd#697]
Compile-time Settings:
  Version: 1.3.6 (stable)
  Platform: FREEBSD11 (FREEBSD11_1) [FreeBSD 11.1-RELEASE-p4 amd64]

Sometime starting around 1.3.6, symlinks with .. in them no longer work. On the server do something like:

```
# mkdir a b
# touch a/file
# cd b
# ln -s ../file link
```

If you look it this from a shell, it looks correct:

`lrwxr-xr-x  1 root  wheel  7 May 13 14:18 link -> ../file`

But if you look at it through proftpd you get:

```
150 Opening ASCII mode data connection for file list
lrw-r-----   1 ftp      ftp             7 May 13 14:18 link -> /file
```

The ".." is missing. You can make the wrong directory completely appear by doing something like:

```
# touch x
# ln -s x y
```

From a shell:

`lrwxr-xr-x  1 root  wheel  1 May 13 14:21 y -> x`

From proftpd:

`lrw-r-----   1 ftp      ftp             1 May 13 14:21 y -> /y/x`

Attempting to access these files fails:

```
ftp> get y
200 EPRT command successful
550 y: No such file or directory
```

Reverting to 1.3.1 fixes this.

**Comment by kdkd** [comment]
This also was [reported on the mailing list](https://sourceforge.net/p/proftp/mailman/message/35795333/) by another user about a year ago:

**Comment by Castaglia** [comment]
I believe this is also the same underlying issue as [Bug#4332](http://bugs.proftpd.org/show_bug.cgi?id=4332), and probably [Bug#4333](http://bugs.proftpd.org/show_bug.cgi?id=4333).

**Comment by Castaglia** [comment]
Would you be willing to see if https://github.com/proftpd/proftpd/pull/801 helps address/change this bad behavior?

**Comment by kdkd** [comment]
Absolutely, I'll let you know when we're able to test it.

**Comment by kdkd** [comment]
That didn't seem to fix it, unfortunately.

```
root@ftpmirror:/z/public/pub/misc/temp # ls
root@ftpmirror:/z/public/pub/misc/temp # mkdir a b
root@ftpmirror:/z/public/pub/misc/temp # touch a/file
root@ftpmirror:/z/public/pub/misc/temp # cd b
root@ftpmirror:/z/public/pub/misc/temp/b # ln -s ../file link
root@ftpmirror:/z/public/pub/misc/temp/b # ls -lash link
1 lrwxr-xr-x  1 root  wheel     7B Apr 23 02:02 link -> ../file
root@ftpmirror:/z/public/pub/misc/temp/b # proftpd --version
ProFTPD Version 1.3.7rc1
```

```
ftp> ls
229 Entering Extended Passive Mode (|||34002|)
150 Opening ASCII mode data connection for file list
lrw-r-----   1 ftp      ftp             7 Apr 23 02:02 link -> /file
```

**Comment by Castaglia** [comment]
How unfortunate.  Thanks for checking; I'll keep working on this.

**Comment by Castaglia** [comment]
This behavior is caused by the changes for [Bug#4219](http://bugs.proftpd.org/show_bug.cgi?id=4219).  As a current workaround, you should be able to use the following in your `proftpd.conf`:
```
<Global>
  ListOptions -a NoAdjustedSymlinks
</Global>
```

**Comment by Castaglia** [comment]
I've now updated PR [https://github.com/proftpd/proftpd/pull/801] to roll back more of the changes done for Bug#4219, if you'd care to try again?

**Comment by kdkd** [comment]
That patch does appear to fix it!

```
ftp> ls
229 Entering Extended Passive Mode (|||17995|)
150 Opening ASCII mode data connection for file list
lrw-r-----   1 ftp      ftp             7 Apr 23 02:02 link -> ../file
226 Transfer complete
ftp> 
```

**Comment by Castaglia** [comment]
🎉 Excellent news! I'll get this merged in, and backported to the 1.3.6 branch, later today.

**Comment by Castaglia** [comment]
PR merged to master, and backported to the 1.3.6 branch; I'm thus closing this ticket.  Thanks!
